### DISCUSSION TITLE
Repository in module always using Master tenant database when accessed in a job defined in the module.

### DISCUSSION THREAD

**Discussion by dkoeder** [oqtane/oqtane.framework#2306] [Q&A] [answered]
Oqtane: current dev branch
Database: LocalDB
Hosting Model: Blazor Server
Database for each Tenant

Created a test module using the Module Management.
Added  TestJob : HostedServiceBase to test module.
Added test module to a page on the Master site and another site.
ExecuteJob in test module, when using repository from test module, will always return data from the Master site's database.

Repositories from the Oqtane framework work fine. The tenant resolution is wrong only on the repository that is in the test module.

I have reviewed #1058, #1061 and #1628, #1637.

I have tried having the repository's DBContext registered by using IService on it and also using an IServerStartup class in module with ConfigureServices services.AddDBContext with options set to ServiceLifetime.Transient with no success. 

Let me know if more info is needed. Thanks.

Edit: Looks like it might work to remove the IService from the Repository class and register the repository from an IServerStartup class with services.AddTransient. Just found this out and want to do some more testing.

**Comment by sbwalker** [comment]
I will have to do some local testing to determine if I can replicate this. I recall that a similar issue was reported in the past however I thought it had been resolved. It is important to note that the metadata and logs for scheduled jobs are stored centrally in the Master database... and scheduled jobs run in the background and have no dependency on the frontend (ie. it does not matter if instances of the module they are packaged with have been added to any sites or not).

The HostedServiceBase contains logic to iterate through the tenants in the installation and set the tenant in the SiteState prior to executing the job:

```
                                        // set tenant and execute job
                                        tenantManager.SetTenant(tenant.TenantId);
                                        notes += ExecuteJob(scope.ServiceProvider);
```

This allows tenant resolution to identify the appropriate tenant database (as HttpContext does not exist in scheduled jobs). Another item which is important to understand is that scheduled jobs do not support standard dependency injection so you need to resolve your services manually within the job:

```
            var siteRepository = provider.GetRequiredService<ISiteRepository>();
```

I will do some further investigation as the default scheduled jobs (ie. Notification and Purge) would suffer from this as well if there was 
an issue.

**Comment by sbwalker** [comment]
I just tested this locally with 2 sites - each with their own isolated database. The Notification job is able to get the information from each database:

![image](https://user-images.githubusercontent.com/4840590/180780400-2ecf3576-ac28-4110-9013-aab0cbddb174.png)

**Reply by dkoeder** [reply]
The repositories, not in modules but in the framework, are added to services with services.AddTransient in OqtaneServiceCollectionExtensions.cs. These repositories work fine in jobs. When a module is created from Module Management the repository has the IService interface. I think any IService is added as Scoped. In my code repositories that are not transient will not do correct tenant resolution in jobs.

**Reply by dkoeder** [reply]
Steps I have found to correct my issue

- remove IService from repository and DBContext in the module
- add an IServerStartup class to module
- in ConfigureServices of IServerStartup do services.AddTransient for repository and  services.AddDbContext for DBContext

**Reply by sbwalker** [reply]
Kudos to you for determining how to utilize the existing extensibility points to solve your problem :) I do believe this can be simplified in the future by introducing an ITransientService interface - I can add this enhancement in 3.1.4.

**Comment by sbwalker** [comment]
@dkoeder thank you for the detailed explanation. I believe you are running into a situation which is similar to #1628 - which was caused  by the DBContext having the wrong scope (it is explained in more detail in that issue thread).

**Comment by sbwalker** [comment]
@dkoeder it appears we need an ITransientService interface in addition to IService (which you are correct only supports Scoped registrations). Note that this problem only arises in scenarios where you want to reference your Repository from an IHostedService - which is more of an edge case which is why it has not been reported until now.

**Comment by sbwalker** [comment, marked as answer]
Added ITransientService in #2311

**Reply by dkoeder** [reply]
@sbwalker Should not line 271 in OqtaneServiceCollectionExtensions.cs be services.AddTransient not services.AddScoped?

                // dynamically register module transient services (ie. server DBContext, repository classes)
                implementationTypes = assembly.GetInterfaces<ITransientService>();
                foreach (var implementationType in implementationTypes)
                {
                    if (implementationType.AssemblyQualifiedName != null)
                    {
                        var serviceType = Type.GetType(implementationType.AssemblyQualifiedName.Replace(implementationType.Name, $"I{implementationType.Name}"));
                        services.AddScoped(serviceType ?? implementationType, implementationType);
                    }
                }

**Reply by sbwalker** [reply]
@dkoeder you are correct - thank you for reviewing. Do you want to submit a PR or do you want me to fix it?

**Reply by dkoeder** [reply]
Fine with me if you fix it.
